src/cascadia/TerminalSettingsModel/WslDistroGenerator.cpp

```diff
@@ -136,3 +136,206 @@ std::vector<Profile> WslDistroGenerator::GenerateProfiles()
+    {
+        auto cbName = maxSubKeyLen + 1;
+        const auto result = RegEnumKeyEx(wslRootKey.get(), i, buffer.get(), &cbName, nullptr, nullptr, nullptr, nullptr);
+        if (result == ERROR_SUCCESS)
+        {
+            guidStrings.emplace_back(buffer.get(), cbName);
+        }
+    }
+    return true;
+}
+
+// Function Description:
+// - Get the list of all the names of all the WSL distros from the registry. If
+//   we fail to open any regkey for the GUID of a distro, we'll just skip it.
+//   Places the names of all the distros into the "names" param.
+// Arguments:
+// - wslRootKey: the HKEY for the Lxss node.
+// - guidStrings: A list of all the GUIDs of the installed distros
+// - names: a vector that recieves all the names of the installed distros.
+// Return Value:
+// - false if the root key was invalid, else true.
+bool _getWslNames(const wil::unique_hkey& wslRootKey,
+                  std::vector<std::wstring>& guidStrings,
+                  std::vector<std::wstring>& names)
+{
+    if (!wslRootKey)
+    {
+        return false;
+    }
+    for (const auto& guid : guidStrings)
+    {
+        wil::unique_hkey distroKey{ _openDistroKey(wslRootKey, guid) };
+        if (!distroKey)
+        {
+            // return false;
+            continue;
+        }
+
+        DWORD bufferCapacity = 0;
+        auto result = RegQueryValueEx(distroKey.get(), RegKeyDistroName, 0, nullptr, nullptr, &bufferCapacity);
+        // request regkey binary buffer capacity only
+        if (result != ERROR_SUCCESS)
+        {
+            // return false;
+            continue;
+        }
+        std::unique_ptr<wchar_t[]> buffer = std::make_unique<wchar_t[]>(bufferCapacity);
+        // request regkey binary content
+        result = RegQueryValueEx(distroKey.get(), RegKeyDistroName, 0, nullptr, reinterpret_cast<BYTE*>(buffer.get()), &bufferCapacity);
+        if (result != ERROR_SUCCESS)
+        {
+            // return false;
+            continue;
+        }
+
+        names.emplace_back(buffer.get());
+    }
+    return true;
+}
+
+// Method Description:
+// - Generate a list of profiles for each on the installed WSL distros. This
+//   will first try to read the installed distros from the registry. If that
+//   fails, we'll fall back to the legacy way of launching WSL.exe to read the
+//   distros from the commandline. Reading the registry is slightly more stable
+//   (see GH#7199, GH#9905), but it is certainly BODGY
+// Arguments:
+// - <none>
+// Return Value:
+// - A list of WSL profiles.
+std::vector<Profile> WslDistroGenerator::GenerateProfiles()
+{
+    static wil::unique_hkey wslRootKey{ _openWslRegKey() };
+    if (wslRootKey)
+    {
+        std::vector<std::wstring> guidStrings{};
+        if (_getWslGuids(wslRootKey, guidStrings))
+        {
+            std::vector<std::wstring> names{};
+            names.reserve(guidStrings.size());
+            if (_getWslNames(wslRootKey, guidStrings, names))

```

Comment:
I believe it would be better to use exception handling as opposed to using booleans.